ui/src/utils/array-utils.ts

```diff
@@ -99,15 +99,15 @@ export function timeStampComparator<T>(
     };
 }
 
-export function multiFieldComparator<T, S>(
-    ...getters: ((obj: T) => S)[]
-): (v1: T, v2: T) => number {
-    return (v1: T, v2: T) => {
-        for (const getter of getters) {
-            const comp = compareByField(v1, v2, getter);
-            if (comp != 0) return comp;
-        }
-        return 0;
+export function multiFieldComparator<T extends unknown[], S>(
+    ...getters: { [K in keyof T]: (obj: T[K]) => S }
+): (v1: T[number], v2: T[number]) => number {
+    return (v1: T[number], v2: T[number]) => {
+        return getters.reduce((previousComparisonResult, nextGetter) => {
+            return previousComparisonResult !== 0
+                ? previousComparisonResult
+                : compareByField(v1, v2, nextGetter);
+        }, 0);
```

Comment:
Tästä tulikin hieman aivojumppaa mietiskellessä. Probleemana oli, että aiemmalla funkkarityypityksellä ei toiminut esimerkiksi käyttötilanne:
```
    multiFieldComparator(
        (entry: { publicationUser: string }) => entry.publicationUser,
        (entry: { publicationTime: TimeStamp }) => entry.publicationTime,
    )
```
koska entryn tyyppi on kaksi eri mahdollista (tai oikeastaan n kpl), ja aiempi funktio oli geneerinen vain T:n osalta niin tämä ei suoraan toimi. Aiempi multifieldComparator ei siis tukenut kuin yhden lähtötyypin T tuolle getterille, kun ylläoleva rakenne olettaa jokaisen getterin lähtötyypin olevan omansa. Tämän voisi aiemmallakin mallilla korjailla käyttämällä samaa tyypitystä kussakin getterissä
```
    multiFieldComparator(
        (entry: PublicationTableItem) => entry.publicationUser,
        (entry: PublicationTableItem) => entry.publicationTime,
    )
```

jolloin jokaiselle eri Entryn-tyypille pitäisi tehdä kuitenkin omat getterit (joka ehkä olisikin selvempää ja näitä samoja gettereitä ei juuri edes käytetä ristiin, julkaisuesikatselussa+julkaisuloki jakaa joitain).

Tuo aiempi sinänsä mielekäs rakenne, että riippumatta entryn varsinaisesta tyypistä, jolla getteriä kutsutaan, voidaan käyttää samoja gettereitä (ja sitä kautta comparatoreja), vaikka entryn tyyppi voi erota getterien eri käyttöpaikoissa (jolloin entry vaan kiinnitetään jokaisessa getterissä sisältämään getterikohtaiset kentät).

Tästä vaan tulee melko ankeaa tyyppispagettia vaikka toimiikin 🤔. Harvemmin tulee nimittäin geneeristä muuttujaa vastaan koodaillessa, joka on N ulotteinen (joka tässä tuo T on). Anekdoottina matikassa tuollainen yleisempi tyypittely onkin sitten taas tyypillisempää.

Reply:
Tällaisia tyypitysongelmia on tullut vastaan aina välillä vuosien varrella, enkä oikein ole törmännyt mihinkään kauhean kivaan tapaan ratkaista sitä. Toinen yleinen ratkaisu tuntuisi olevan tehdä X kappaletta eri tyyppioverrideja, jotka menevät tyyliin `func<T>`,  `func<T, T2>`, `func<T, T2, T3>`... Se tuntuisi tosin todella tönköltä ratkaisulta, varsinkin meille tähän käyttötarkoitukseen. On tuo ehdotettu getter-ratkaisukin vähän tönkkö, mutta ei kyllä yhtä tönkkö, joten ehkä se on parasta mitä tähän maailmanaikaan on tarjolla.